rust_src/src/data.rs

```diff
@@ -789,4 +790,36 @@ pub fn logcount(value: EmacsInt) -> i32 {
     unsafe { rust_count_one_bits(value as usize) }
 }
 
+/// Parse STRING as a decimal number and return the number.
+/// Ignore leading spaces and tabs, and all trailing chars.  Return 0 if
+/// STRING cannot be parsed as an integer or floating point number.
+///
+/// If BASE, interpret STRING as a number in that base.  If BASE isn't
+/// present, base 10 is used.  BASE must be between 2 and 16 (inclusive).
+/// If the base used is not 10, STRING is always parsed as an integer.
+#[lisp_fn(min = "1", name = "string-to-number", c_name = "string_to_number")]
+pub fn string_to_number_lisp(mut string: LispStringRef, base: Option<EmacsInt>) -> LispObject {
+    let b = match base {
+        None => 10,
+        Some(n) => {
+            if !(n >= 2 && n <= 16) {
+                xsignal!(Qargs_out_of_range, base.into());
+            }
+            n
+        }
+    };
+
+    let p = string.sdata_ptr();
+    unsafe {
+        while *p == ' ' as c_char || *p == '\t' as c_char {
+            *p += 1;
```

Comment:
This is your bug. The pointer is not moved deeper into the string. Instead, if the character is whitespace you increment the _character_ by one. The first sign you had this wrong is that `p` was not `mut`. If you are moving the pointer it ought to be mutable :-)

In Rust, to do pointer math you have to use `offset`.

```
-    let p = string.sdata_ptr();
+    let mut p = string.sdata_ptr();
     unsafe {
         while *p == ' ' as c_char || *p == '\t' as c_char {
-            *p += 1;
+            p = p.offset(1);
         }
     }
```

Reply:
Thanks. I forgot about the offset stuff. I already made a PR using offsets last year and it took some time until I solved this exact problem ;)
Any idea why this doesn't work ? I mean this is an unsafe block, so why can't rust behave like C in this situation.

Reply:
This was a conscious design decision by the Rust team to avoid this exact kind of Rust bug. In C you get:

`*p++ = foo;` and `p++ = foo;` mixed up quite often.